doc/negraimport.1
.\" Man page for negraimport
.\" Use the following command to view man page:
.\"
.\"  tbl negraimport.1 | nroff -man | less
.\"
.TH NEGRAIMPORT 1 "January 20, 2007"
.SH NAME
negraimport \- A tool to convert NEGRA format to Emdros MQL
.SH SYNOPSIS
\fBnegraimport\fR [ options ] [input_filename ...]
.br
.SH DESCRIPTION
\fBnegraimport\fR is a command-line tool to convert NEGRA
format to Emdros MQL for later importing into Emdros.

.SH OPTIONS
\fBnegraimport\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help, then quit
.TP
.BI \-V\ ,\ \-\-version
show version, then quit
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".
.TP
.BI \-s\ ,\ \-\-schema
show MQL schema on stdout, then quit (can be used with -d)
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set database name. If used with -s, the string "CREATE DATABASE
'dbname' GO USE DATABASE 'dbname' GO" will be issued before the schema.  If used when importing, "USE DATABASE 'dbname' GO" will be issued before anything else.
.TP
.BI \-o\ ,\ \-\-output \ filename
dump to file filename. The default is "-", which means "standard
output".
.TP
.BI \-\-start-monad \ monad
The start monad to use. Must be >= 1.  The default is 1.
.TP
.BI \-\-start-id_d \ id_d
The start id_d to use. Must be >= 1.  The default is 1.
.TP

.SH OPERATION

\fBnegraimport\fR reads corpus data in NEGRA format and converts the
data to MQL statements for later importing into Emdros.

The filenames given after the options on the command line are
interpreted as if each of them contains one NEGRA document, each
containing one or more trees.  If no filenames are given, the input is
read from stdin.

If no \fB-o\fR switch is given, the output is printed on stdout.

If an error occurs, the string "FAILURE" or the string "ERROR" is
printed on stderr, along with an error message.

If no error occurs, a string of the form "SUCCESS: next_monad is X
next_id_d is Y" is printed on stderr, where X and Y are positive
integers denoting the next monad and the next id_d to be used by the
next invocation of the program, respectively.  This is useful if
you've got several directories' worth of documents to import.

.SH SCHEMA

The schema can be seen by giving the program \fB-s\fR switch, with an
optional \fB-d\fR switch.

A "Sentence" corresponds to one top-level sentence.

A "Phrase" is a nonterminal in the tree which is not a top-level
sentence.  Note that this encompasses clauses ("S") as well as
phrases.  Its "parent" feature points to its parent.  Its parents
feature feature gives a list of id_ds of all its parents, right up to
the "Sentence" node.  Its "tag", "morphtag", and "edge" features are
what you think they are.  There are five pairs of features "secedgeX"
and "secparentX", for X in the range 1 to 5 (inclusive).  These
features show secondary edges.

A "Word" is a terminal node in the tree (i.e., a token). Its features
are largely the same as for "Phrase", except that it has a "surface"
feature which gives the surface of the token, as well as a "postag"
instead of a "tag".


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (internal error)"
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.


- *** Version 3.6.1 *** 

Release notes for Emdros 3.6.1
==============================

Emdros version 3.6.1 is a maintenance release with the following
summary of changes since 3.6.0:

- The MQL query language has been extended with the ability to compare
  features with type LIST OF INTEGER, LIST OF ID_D, and LIST OF
  <enum-type> with an explicit list (including the empty list) using
  the "=" comparison operator.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause_atom
     daughters = () // Test for empty list
     OR
     daughters = (12345,67891) // Test for explicit, non-empty list
  ]
  GO

- Detection of PHP 5 and PHP 7 on macOS now works correctly.

- Compileability on Solaris 10 with Sun's CC C++ compiler has been
  restored.

- Compatibility with wxWidget 2.8.12 has been restored.


Enjoy!


Ulrik Sandborg-Petersen



- *** Version 3.6.0 *** 

Release notes for Emdros 3.6.0
==============================

The release notes below cover some of the important changes in Emdros
since 3.5.0.

- SQLite2 support has been removed, as version 2 of SQLite has not
  been updated by the SQLite authors for many, many years.  Also,
  SQLite2 support in Emdros has been been deprecated for some time.

- The MQL subsystem has a better query planner, which is able to
  choose the "Innermost First" strategy, rather than the default
  "Outermost First" strategy, under many more circumstances.

- The Emdros MySQL backend now has support for transactions, if
  available in the MySQL server.

- Assorted performance improvements to the following subsystems:

  - XML parsing (renderobjects, for example)
  - MQL

- Node.JS support has been removed.  The reason is that the latest
  SWIG version (3.0.12) does not support the latest Node.  Also, the
  Node.JS bindings never really worked.

- PCRE has been upgraded.

- SQLite3 has been upgraded.

- JSON output now escapes string characters correctly.

- Compiling Linux .rpm packages works on recent systems (CentOS 7,
  Fedora 28).

- Compiling Linux .deb packages works on recent systems (Ubuntu 16.04,
  Ubuntu 18.04, Debian 9.4).

- Building on Windows supports the latest Visual Studio.

- Lots of compiler warnings have been removed.

- the harvest template language has new constructs:

  - {{ inccounter <string> var <string> }}
  - {{ deccounter <string> var <string> }}

- The Full Text Search FTS3 subsystem is now able to obey
  largest_proximity, and to use * and ? in the query terms.

- A new program was added, bptdetach(1), which enables detaching
  "payloads" from BPT databases.

- The BPT backend (if available -- ask Ulrik for the source code
  licensing terms if you are interested) now supports LZMA compression
  of payloads.

- Source code using wxWidgets has been updated to support wxWidgets
  2.8, 3.0, and 3.1.

- The Emdros PostgreSQL backend now adds "TEMPLATE template0" when
  doing CREATE DATABASE.  This is so as to avoid using template1,
  which may conflict with the encoding used (and does so on Debian
  9.4).

- The amalgamation output has been split into several files.  This is
  because the Android NDK MIPS target cannot compile the old,
  monolithic amalgamation files.



- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries

src/negraimport.cpp
/*
 * negraimport.cpp
 *
 * A front-end to the Negra import
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 5/11-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */



#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc

#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <emdros-lconfig.h>
#include <emdfdb.h>

#include <negraimporter.h>

#include <opt.h>




int exec_list(const std::list<std::string>& input_filename_list, std::ostream* pOutStream, const std::string& dbname, monad_m start_monad, id_d_t start_id_d, bool bPrintSchema)
{
	int result = 0;

	NEGRAImporter *pImporter = new NEGRAImporter(start_monad, start_id_d);

	if (bPrintSchema) {
		if (dbname != "") {
			std::cout << "CREATE DATABASE '" << dbname << "' GO\n\n"
				  << "USE DATABASE '" << dbname << "' GO\n\n";
		}
		
		pImporter->putSchema(&std::cout);

		delete pImporter;

		return 0;
	}


	bool bResult = false;
	if (input_filename_list.empty()) {
		bResult = pImporter->readDocumentFromStream(&std::cin);
	} else {
		bResult = pImporter->readDocumentsFromFiles(input_filename_list);
	}

	if (!bResult) {
		std::cerr << "FAILURE: Error occurred reading documents." << std::endl;
		result = 5;
	} else {
		if (dbname != "") {
			(*pOutStream) << "USE DATABASE '" << dbname << "' GO\n\n";
		}

		pImporter->putCorpusAsMQL(pOutStream);

		std::cerr << "SUCCESS: next_monad is " << pImporter->getNextMonad() << "   next_id_d is " << pImporter->getNextID_D() << std::endl;
	}

	delete pImporter;

	return result;
}

bool open_file_for_writing(std::ofstream& fout, const std::string& filename)
{
	bool bResult = false;
	fout.open(filename.c_str());
	if (!fout) {
		std::cerr << "FAILURE: Could not open file " << filename << " for writing." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}

bool open_file_for_reading(std::ifstream& fin, const std::string& filename)
{
	bool bResult = false;
	fin.open(filename.c_str());
	if (!fin) {
		std::cerr << "FAILURE: Could not open file " << filename << " for reading." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}

void show_version(std::ostream& ostr)
{
	ostr << "negraimport from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   negraimport [options] [inputfile ...]" << std::endl;
	ostr << "   negraimport (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	ostr << "   --help               Show this help" << std::endl;
	ostr << "   -V , --version       Show version" << std::endl;
	ostr << "   -s , --schema        Print schema and quit." << std::endl;
	ostr << "   -d , --dbname db     database name (to use in MQL output)" << std::endl;
	ostr << "   -o , --output file   Dump database to filename file (\"-\" means stdout)" << std::endl;
	ostr << "   --start-monad monad  Start monad (default 1)" << std::endl;
	ostr << "   --start-id_d id_d    Start id_d (default 1)" << std::endl;
	ostr << "DEFAULTS:" << std::endl;
	ostr << "   -o -                 (Use stdout)" << std::endl;
	ostr << "   --start-monad 1" << std::endl;
	ostr << "   --start-id_d  1" << std::endl;
	ostr << "   If no filenames are given, a document is read from stdin." << std::endl;
	ostr << "   By default, no -d switch is assumed." << std::endl;
}

int main(int argc, char* argv[])
{
	// Set defaults
	std::string db;
	std::string output_filename("-");
	std::list<std::string> input_filename_list;
	bool bPrintSchema = false;
	monad_m start_monad = 1;
	id_d_t start_id_d = 1;
	std::string dummy_hostname, dummy_user, dummy_password;
	eCharsets dummy_charset;
	eBackendKind dummy_backend_kind;

	addStandardArguments(false // Do not add -e | --encoding
			     );

	addOption("-s", "--schema");

	addOption("-d", "--dbname", true, 
		  "", // Default value
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");

	addOption("-o", "--output", true, 
		  "-", // Default value
		  "ERROR: -o and --output must have a filename (or \"-\" for stdout)\n"
		  "       as their argument.\n"
		  "       example: -o output.xml\n");


	addOption("--start-monad", "--start-monad", true, 
		  "1", // Default value
		  "ERROR: --start-monad must have a monad as its argument\n"
		  "       example: --start-monad 1\n");

	addOption("--start-id_d", "--start-id_d", true, 
		  "1", // Default value
		  "ERROR: --start-id_d must have an id_d as its argument\n"
		  "       example: --start-id_d 1\n");


	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  dummy_hostname,
					  dummy_user,
					  dummy_password,
					  dummy_backend_kind,
					  dummy_charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// --start-monad
		std::string start_string;
		if (getArgumentValue("--start-monad", start_string)) {
			if (string_is_number(start_string)) {
				start_monad = string2long(start_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --start-monad must be a number, and not '" << start_string << "'." << std::endl;
				return 1;
			}
		}
		if (start_monad < 1) {
			print_usage(std::cerr);
			std::cerr << "ERROR: The argument of --start-monad must be >= 1." << std::endl;
			return 1;
		}

		// --start-id_d
		if (getArgumentValue("--start-id_d", start_string)) {
			if (string_is_number(start_string)) {
				start_id_d = string2long(start_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --start-id_d must be a number, and not '" << start_string << "'." << std::endl;
				return 1;
			}
		}
		if (start_id_d < 1) {
			print_usage(std::cerr);
			std::cerr << "ERROR: The argument of --start-id_d must be >= 1." << std::endl;
			return 1;
		}

		// --dbname | -d
		getArgumentValue("-d", db);

		// Output
		getArgumentValue("-o", output_filename);

		// -s | --schema
		if (getArgumentPresent("-s")) {
			bPrintSchema = true;
		}

		// Any and all surplus arguments are input filenames...
		input_filename_list = surplus_arguments;
	}


	int nResult = 3;
	bool bContinue = true;
	try {
		std::ostream *pOut = 0;
		std::ofstream fout;
		if (output_filename == "-") {
			pOut = &std::cout;
		} else {
			if (!open_file_for_writing(fout, output_filename)) {
				nResult = 4;
				bContinue = false;
			} else {
				pOut = &fout;
			}
		}

		if (bContinue) {
			nResult = exec_list(input_filename_list, pOut, db, start_monad, start_id_d, bPrintSchema);
		}
	} catch (EMdFNULLValueException& e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError& e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	}

	// Clean up
	//	delete pEE;
	return nResult;
}


